pallets/gas/src/lib.rs

```diff
@@ -193,85 +255,43 @@ pub mod pallet {
             Allowance::<T>::mutate(|v| *v = v.saturating_sub(gas));
         }
 
-        /// Check if a node is consumed and does not have any child nodes so it can be deleted.
+        /// Performs, if possible, cascade deletion of multiple nodes on the same path from the node with `key` id
+        /// to the tree's root.
+        ///
+        /// There are two requirements for the node to be deleted:
+        /// 1. Marked as consumed.
+        /// 2. Has no child refs.
+        ///
         /// If the node's type is `ValueType::External`, the locked value is released to the owner.
-        /// Otherwise this function is called for the parent node to propagate the process further.
+        /// Otherwise pre-delete ops are executed, then the node is deleted and after that the same procedure
+        /// is repeated on the node's parent until it's marked consumed and has no child refs.
+        ///
+        /// Pre-delete ops are:
+        /// 1. Parents refs decrease.
+        /// 2. Value movement to the first parent, which can hold specified value.
+        ///
+        /// The latter op is required, because node can be marked as consumed, but still has non-zero inner value.
+        /// That is the case, when node was splitted without gas and then consumed. So when node's gas-less child
+        /// is consumed the [`Self::check_consumed`] function is called on the consumed parent with non-zero value.
         pub(super) fn check_consumed(key: H256) -> Result<ConsumeOutput<T>, DispatchError> {
-            let mut delete_current_node = false;
-            let maybe_node = Self::get_node(key);
-            let outcome = if let Some(current_node) = maybe_node {
-                match current_node.inner {
+            let mut node = Self::get_node(key).ok_or(Error::<T>::NodeNotFound)?;
+            while node.consumed && node.refs() == 0 {
+                node.decrease_parents_ref::<T>()?;
+                node.move_value_upstream::<T>()?;
+                GasTree::<T>::remove(node.id);
+
+                match node.inner {
+                    ValueType::External { id, value } => {
+                        return Ok(Some((NegativeImbalance::new(value), id)))
+                    }
```

Comment:
Isn't this PR a best one to apply old considering stuff? I want to provide gas unlocking directly to user for cases when node.inner is not only External one.

EXTERNAL_NODE (consumed, 2 refs left somewhere)
FIRST_NODE (consuming, no refs left after SECOND_NODE consume)
SECOND_NODE (consuming, no refs)

SECOND_NODE on consume moves gas upper to FIRST_NODE and upper to EXTERNAL_NODE

due to the fact that EXTERNAL_NODE is already consumed, this seems logic to repatriate that value back to user
probably this logic may be only useful if no unspec refs left

I suggest to discuss all pros and cons with the team on sync up

Reply:
discussed, will be implemented in the separate PR after some research and discussions

Reply:
https://github.com/gear-tech/gear/issues/1072